### ISSUE TITLE
Discussion: syntax for plotting from a DataFrame

### ISSUE THREAD

**Issue by piever** [JuliaPlots/StatPlots.jl#72]
Issues with the current syntax for plotting from a DataFrame:
- It's implemented as a user recipe, which means it doesn't play well with grouping (see [#1033](https://github.com/JuliaPlots/Plots.jl/pull/1033))
- The user may be led to expecting that it is equivalent to typing the columns inside the DataFrame (which I believe it should be), but instead it is different in subtle ways, for example when trying to group by more than one column. This fails

```
using StatPlots, DataFrames, RDatasets
iris = RDatasets.dataset("datasets","iris")
iris[:LongSepal] = iris[:SepalLength] .> median(iris[:SepalLength])
scatter(iris, :PetalLength, :PetalWidth, group = collect(zip(:Species,:LongSepal)))
```

but it'd work with the columns:

`scatter(iris[:PetalLength], iris[:PetalWidth], group = collect(zip(iris[:Species],iris[:LongSepal])))`

- other recipes also fail because the transformation from symbol to columns happens to late (see #59 ):

`corrplot(iris, [:SepalLength :SepalWidth :PetalLength :PetalWidth])`

- it has to resort to heuristics to decide whether a symbol is a symbol or a column. For example 
`scatter(iris, :PetalLength, :PetalWidth, color = :blue)` plots in blue, unless there is a column in `iris` called `:blue`

- it has to resort to a runtime `eval` call

My personal preference would be for syntactically replacing symbols with columns.
Possible options:

- Adding a dispatch to `plot` if the first argument is a DataFrame: in that case all symbols and expression are replaced by the respective columns, using the same functions we use now. It wouldn't resolve the ambiguity symbol/column + calling `eval` but at least wouldn't mess with the recipe system

- Officialize the `@with` method from DataFramesMeta (a bit verbose but solves all of the above problems: `group` and `corrplot` work just fine and one can escape symbols to prevent them from becoming columns). Extra advantage: it can be used in blocks:

```
@with iris begin
    plt = plot(:SepalLength,:PetalWidth)
    plot!(:SepalLength,:PetalLength)
    display(plt)
end
```

- A slight modification of `@with` where each symbol `s` is not replaced by `df[s]`, but by 
`haskey(df, s) ? df[s] : s` so that the user wouldn't need to escape symbols so often
 
- Try and find a less verbose macro, though the implementation is not super straightforward. One bold possibility would be sth like:

```
macro plot(args...)
    :(@with $(args[1]) $(Expr(:call, :plot, Base.tail(args)...)))
end
@plot(iris, :PetalLength, :PetalWidth)
```

But it seems hard to get it to work with keyword arguments/ different seriestypes.

**Comment by piever** [comment]
To make the discussion more complete, here is an example of a simplified `@with` macro called `@given`:

```
macro given(d, x)
    esc(_given(d,x))
end

_given(d, x) = x

function _given(d, x::Expr)
    (x.head == :quote) && return :(haskey($d, $x) ? $d[$x] : $x)
    (x.head == :call) && x.args[1] == :^ && length(x.args) == 2 && return x.args[2]
    return Expr(x.head, _given.(d, x.args)...)
end
```

Usage:

```
using StatPlots, DataFrames,RDatasets
iris = RDatasets.dataset("datasets","iris");
@given iris scatter(:PetalLength, :PetalWidth, group = :Species, legend = :topleft)
iris[:topleft] = rand(150)
@given iris scatter(:SepalLength, :SepalWidth, group = :Species, legend = ^(Symbol("topleft")))
```

Advantages over `@with` : 
- No need to escape the symbols unless there actually is ambiguity
- No DataFramesMeta dependency
- It's easy to adjust it to include other datasources, provided they allow extracting one column (which I'm assuming all of them do)

**Comment by piever** [comment]
I am sure I had written this already here, but apparently I forgot to post it:  as a first non-breaking step,  if there is consensus, I'd be happy to make a PR with the `@given` macro, mentioning it instead of `@with` in the README. Of course we should maybe spend some time figuring out what's the best name for `@given`.

Even if we get the `plot(df,...)` syntax to work, it's good to also have the macro version because it retains some advantages that are just impossible to achieve with a function:

- Manipulating columns doesn't require wrapping everything in a `:()`. For example one can write 
```julia
@given iris scatter(log.(:PetalLength), :PetalWidth .+ 1, group = :SepalWidth .> 3, legend = :topleft)
```
and it would just work (I'm even tempted to add a `where` kw to also select what data to plot in the plot command)
- Manipulating columns doesn't require a runtime `eval`

**Comment by mkborregaard** [comment]
Sorry for not responding - I'm reading and thinking carefully but haven't had a chance to do it justice - please have a little patience :-)

**Comment by kescobo** [comment]
My gut reaction is that adding additional syntax would be unfortunate given the goals of the `Plots.jl` ecosystem. Then again, the functionality is more important, and if this would fix some of the issues with using DataFrames to their full extent... 

Alas, I don't understand enough of the backend stuff to comment on implementation.

**Comment by mkborregaard** [comment]
I think this looks really cool. It's certainly a lot simpler than the current non-macro framework!! If you'd make a PR with the given macro I'll try it out in some of the other cases I've been considering. This involves e.g. passing a part of a DataFrame as a matrix but that should be possible with `[:a :b]` here I think.

I think we should be sure to be compatible not just with DataFrames but with abstract table formats such as DataTables, JuliaDB, IndexedTables etc but I'll need to check the current status of those to see if that could work.

Another thing I'd like the macro to do was to replace all NAs with NaNs for number fields, and "" / Symbol() for string/symbol fields, so that the plot function doesn't automatically fail when passing it a DataFrame with NAs in it.

In short, I agree that a macro approach is the best way to deal with table symbols.

**Comment by piever** [comment]
Your intuition is correct, using  the horizontal concatenation syntax:
`@given iris corrplot([:SepalLength :SepalWidth :PetalLength :PetalWidth])`
![corrplot](https://user-images.githubusercontent.com/6333339/29634684-8aa51214-8842-11e7-8365-dfcaf9324612.png)

I'll add the missing value part and make a PR tomorrow. 

As for the support of any table-like structure, maybe the easiest is to support `IterableTables`? I'm also not sure what the state of affairs is in the data echosystem...

**Comment by mkborregaard** [comment]
This should be fully resolved now.
